OsmAnd/src/net/osmand/plus/smartnaviwatch/SmartNaviWatchPlugin.java

```diff
@@ -0,0 +1,501 @@
+                distanceToNextPoint = MapUtils.getDistance(lastKnownLocation.getLatitude(), lastKnownLocation.getLongitude(), p2.getLatitude(), p2.getLongitude());
+            }
+            double visibleRange = Math.min(maxVisibleRange, Math.max(minVisibleRange, distanceToNextPoint + mapBorderSpacing));
+
+            BinaryMapIndexReader.SearchRequest<BinaryMapDataObject> request = buildRequestAround(lastKnownLocation, visibleRange);
+
+            List<BinaryMapDataObject> res = null;
+            try {
+                 res = reader.searchMapIndex(request);
+            } catch(IOException ex) {}
+
+            if (res != null){
+                // Set the users position and the current view range
+                c.setUserPosition(new PolygonPoint(MapUtils.get31TileNumberX(lastKnownLocation.getLongitude()), MapUtils.get31TileNumberY(lastKnownLocation.getLatitude())));
+                c.setTopLeftViewRange(new PolygonPoint(request.getLeft(), request.getTop()));
+                c.setBottomRightViewRange(new PolygonPoint(request.getRight(), request.getBottom()));
+
+                //set the bearing only if the appropriate setting is set
+                if (application.getSettings().ROTATE_MAP.get() == OsmandSettings.ROTATE_MAP_BEARING) {
+                    // set only when speed is fast enough
+                    if (lastKnownLocation.getSpeed() >= 0.5) {
+                        c.setRotation(-lastKnownLocation.getBearing());
+                    }
+                }
+
+
+                // Add all the objects on the map
+                for(BinaryMapDataObject o : res) {
+                    MapPolygon poly = polygonFromDataObject(o);
+                    if (poly.getType() != MapPolygonTypes.UNKNOWN) c.add(poly);
+                }
+
+                // Add the current routing path if found
+                if (currentInfo != null) {
+                    List<Location> routePoints = routing.getRoute().getImmutableAllLocations();
+                    PolygonPoint[] pointPath = new PolygonPoint[routePoints.size()];
+                    PolygonPoint[][] emptyInner = new PolygonPoint[0][];
+
+                    for(int rp = 0; rp < routePoints.size(); rp++){
+                        Location l = routePoints.get(rp);
+                        pointPath[rp] = new PolygonPoint(MapUtils.get31TileNumberX(l.getLongitude()), MapUtils.get31TileNumberY(l.getLatitude()));
+                    }
+
+                    c.add(new MapPolygon(MapPolygonTypes.ROUTE_PATH, pointPath, emptyInner));
+                }
+
+                Collections.sort(c.getPolygons());
+                c.normalize();
+
+
+            }
+        }
+
+        return c;
+    }
+
+    /**
+     * Converts a MapDataObject into a polygon
+     * @param obj    Map Data Object
+     * @return       Polygon
+     */
+    private MapPolygon polygonFromDataObject(BinaryMapDataObject obj) {
```

Comment:
Is it simplified rendering for watch? Did you consider to create special rendering style for that.